### PR TITLE
Ticket 4143: Knauer 1050 Update OPI

### PR DESCRIPTION
**Pull request by ghost**
Added error handling and minor bug fixes

### CODE REVIEW

File: knr1050Sup/knr1050.proto [hunk 1]
```diff
@@ -35,13 +35,23 @@ get_status {
...
+    @init{get_mode;}
+}
+
+resetError {
+    out "RESET_ERROR";
+    in "OK";
 }
+
+readErrorString {
+    in "ERROR:%#s";
```
Comment (Alistair-McGann-Tessella) [inline knr1050Sup/knr1050.proto:55]: This should be called from an exception handler. At the minute this causes a mismatch to be raised by the `$(P)PRESSURE:LIMITS` PV, which can be seen in the IOC output
Documentation:
http://epics.web.psi.ch/software/streamdevice/doc/protocol.html#except

File: knr1050Sup/knr1050.db [hunk 2]
```diff
@@ -656,6 +670,20 @@ record(bi, "$(P)ERR_IN") {
     info(INTEREST, "HIGH")
 }
 
+record(stringin, "$(P)ERROR:STR") {
+    field(DESC, "Input error return")
+	field(DTYP, "stream")
+	field(SCAN, "I/O Intr")
```
Comment (Alistair-McGann-Tessella) [inline knr1050Sup/knr1050.db]: Please use spaces instead of tabs for whitespace

**Review by Alistair-McGann-Tessella — CHANGES_REQUESTED** [review]

File: knr1050Sup/knr1050.db (hunk 2 above)
Comment (ghost) [reply knr1050Sup/knr1050.db]: Converted tabs to spaces.

File: knr1050Sup/knr1050.proto (hunk 1 above)
Comment (ghost) [reply knr1050Sup/knr1050.proto:55]: This device can return different error strings depending on its operating mode. We want to capture what these error returns are and allow the user to read them for troubleshooting (they are suppressed using a mismatch handler), as these can relate to input errors, status modes, device failures (seal etc).

Originally I attempted to create mismatch handlers for each record, but they got fairly complex. Tom sat down with me and we can up with this general solution (that is also used on the JASCO4180).

I think it works sufficiently without adding additional complexity.

Comment (Alistair-McGann-Tessella) [reply knr1050Sup/knr1050.proto:55]: :+1: I'm happy this is a good solution

**Pull request merged**